saleor/checkout/complete_checkout.py

```diff
@@ -1053,21 +1053,33 @@ def _handle_allocations_of_order_lines(
     )
 
 
-def _handle_checkout_discount(order: "Order", checkout: "Checkout"):
+def _create_order_discount(order: "Order", checkout_info: "CheckoutInfo"):
+    checkout = checkout_info.checkout
     if checkout.discount:
-        # store voucher as a fixed value as it this the simplest solution for now.
-        # This will be solved when we refactor the voucher logic to use .discounts
-        # relations
-
-        order.discounts.create(
-            type=DiscountType.VOUCHER,
-            value_type=DiscountValueType.FIXED,
-            value=checkout.discount.amount,
-            name=checkout.discount_name,
-            translated_name=checkout.translated_discount_name,
-            currency=checkout.currency,
-            amount_value=checkout.discount_amount,
-        )
+        checkout_discount = checkout.discounts.first()
+        if checkout_discount and checkout_discount.type == DiscountType.ORDER_PROMOTION:
```

Comment:
Instead of two `if`'s, where bot of them create the `OrderDiscount`. Can't we make a single point of calling `order.discounts.create`, and above this, have a logic that will prepare the data?
Something like this:
```python
    if checkout.discount:
        order_discou_data = {}
        if voucher:
            order_discout_data = {...}
        elif checkout_discount.type == DiscountType.ORDER_PROMOTION:
            order_discout_data = {...}
        # create OrderDiscount
```

Reply:
I was just thinking about having a place to build the payload for `OrderDiscount`, and one place for calling `order.discount.create`. It is a cosmetic detail, but IMHO it is simpler to read the code. 
Something like this:
```python

def _get_data_for_order_discount(checkout_info: "CheckoutInfo"):
    """Prepare order discount data based on discounts assigned to Checkout"""

    checkout_discount = checkout_info.checkout.discounts.first()

    if checkout_discount and checkout_discount.type == DiscountType.ORDER_PROMOTION:
        data = model_to_dict(checkout_discount, fields=[
            "type",
            "value_type",
            "value",
            "name",
            "translated_name",
            "currency",
            "amount_value",
        ])
        data["promotion_rule"] = checkout_discount.promotion_rule
        return data

    # Is it correct? Or maybe it should be `not checkout_discount`
    if checkout_info.voucher:
        # return the data based on the voucher data
        return {
            "type" :DiscountType.VOUCHER,
            "value_type" : DiscountValueType.FIXED,
            "value" : checkout_info.checkout.discount.amount,
            "name" : checkout_info.checkout.discount_name,
            "translated_name" : checkout_info.checkout.translated_discount_name,
            "currency" : checkout_info.checkout.currency,
            "amount_value" : checkout_info.checkout.discount_amount,
            "voucher" : checkout_info.voucher,
            "voucher_code" : checkout_info.voucher_code.code
            if checkout_info.voucher_code
            else None,
        }




def _create_order_discount(order: "Order", checkout_info: "CheckoutInfo"):
    checkout = checkout_info.checkout
    if not checkout.discount:
        return None
    discount_data = _get_data_for_order_discount(checkout_info)
    if not discount_data:
        return None
    order.discounts.create(**discount_data)
```
But this is only a suggestion.